[Parse][Sema] Improve interpolation parsing and construction

beccadax opened the pull request:
In Swift 4.2 mode, we currently diagnose and correct multi-argument and labeled interpolations during parsing. This isn't great because there are various modes where we parse code that won't actually compile in Swift 4.2 mode, like code in `#if swift(>=5)` blocks.

This PR moves this diagnosis to Sema, specifically to `PreCheckExpression`. This then allows us to parse string interpolations *literally* as argument lists, which (with another small fix) is enough to let code completion offer `appendInterpolation` overload signatures in a string interpolation segment.

While I was there, I also cleaned up some issues with the strange interpolation warning test, such as fixits that weren't being checked properly and accidental indentation. It's best to review the commits separately so you can see which parts of the test changed.

Resolves [SR-9937](https://bugs.swift.org/browse/SR-9937).

beccadax committed:
Improve strange-interpolation test

There were a number of mistakes in this test:

• The whole thing was indented one space.
• Some fix-it tests were malformed and therefore not being tested.
• The output checking could in theory allow content before or after the intended content.

beccadax committed:
Move strange interpolation fix to PreCheckExpression

This defers diagnosis until a stage where #if conditions have definitely been evaluated, at the cost of a slightly more complex implementation. We’ll gain some of that complexity back in a subsequent refactoring. Fixes SR-9937.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test

rintaro commented:
Thank you for looking into this!
Instead of guarding the block in Parser, is it possible to move this diagnostics and transformation to Sema?

beccadax committed:
[Parse] Parse string interpolations as args

Now that we manipulate the argument list to correct strange interpolations in Sema, we can parse interpolations directly as argument lists, simplifying the parser.

By itself, this refactoring causes a code completion regression; a subsequent commit will fix that.

brentdax commented:
@rintaro I've moved it to `PreCheckExpression`, which runs just before type checking, and modified the tests to simply check that we don't diagnose in a failing `#if` in parse mode without checking that we *do* diagnose outside of one. However, the obvious cleanup this enables—parsing string interpolations by calling `parseExprCallSuffix()`—changes code completion behavior in a way that mostly seems to be an improvement, [but has some problems](https://github.com/apple/swift/pull/24464/commits/d2539990c59c5cd4412c62a4ea5f18453125c01d#diff-ef8965118aeed7d38e0eb83231430b04). Do you have any thoughts about how to fix the little bugs mentioned in the FIXMEs?

rintaro commented:
`TypeRelation[Invalid]` probably because [the context type analysis in code completion](https://github.com/apple/swift/blob/7311886219a5c636b27d913654b3d47bac918a4c/lib/IDE/ExprContextAnalysis.cpp#L685) is not working correctly. I need to look AST dump to investigate it.

beccadax committed:
[AST] Give appendInterpolation refs a SourceLoc

This change permits UnresolvedDotExpr to have both a name and a base that are implicit, but a valid DotLoc, and to treat that DotLoc as the node’s location. It then changes the generation of string interpolation code so that `$stringInterpolation.appendInterpolation` references have a DotLoc corresponding to the backslash in the string literal.

This makes it possible for `ExprContextAnalyzer` in IDE to correctly detect when you are code-completing in a string interpolation and treat it as an `appendInterpolation` call.

brentdax commented:
@swift-ci please smoke test

rintaro commented:
Ah, I see that was caused by invalid `SourceLoc`s. Thank you!

rintaro commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -952,6 +952,104 @@ namespace {
+
+        virtual std::pair<bool, Expr *> walkToExprPre(Expr *E) {
+          assert(!isa<InterpolatedStringLiteralExpr>(E) &&
+                 "StrangeInterpolationRewriter found nested interpolation?");
```
Could you add some nested test cases like:
```swift
"[\(foo: "=\(bar: 0)=")]"
```

brentdax replied:
Will do. (But it's safe because the only place nested interpolations can appear is as one of the parameters of an `appendInterpolation(_:)` call, and we stop recursing when we see `CallExpr`.)

brentdax replied:
Done.

brentdax commented:
@swift-ci please build toolchain

brentdax commented:
So…I accidentally made `appendInterpolation(…)` overloads autocomplete their parameter lists correctly.

Cool.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
Squashed away the intermediate state.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test and merge
